src/steamship/agents/llms/steamship_llm.py

```diff
@@ -0,0 +1,110 @@
+        :param tools: A list of Tools, representing functions which can be called with this LLM. Providing this requests
+          the FunctionCallingSupport capability, and provides the function definitions for it.
+        :param assert_capabilities: If True (default), raise a SteamshipError if the LLM plugin did not respond with a
+          block that asserts what level capabilities were fulfilled at.
+        :param kwargs: Options that can be passed to the LLM model as other parameters.
+        :return: a List of Blocks that are returned from the plugin.
+        """
+
+        # TODO (PR callout): I'm not certain this class needs to be abstract?  This seems pretty soup-to-nuts.
+        capabilities = []
+        blocks = messages[:]
+        if system_prompt:
+            blocks += system_prompt
+            capabilities.append(SystemPromptSupport())
+
+        if history:
+            blocks += history
+            capabilities.append(ConversationSupport())
+
+        if tools:
+            capabilities.append(
+                FunctionCallingSupport(functions=[tool.as_openai_function() for tool in tools])
+            )
+
+        blocks.append(CapabilityPluginRequest(requested_capabilities=capabilities).to_block())
+        # TODO (PR callout): I have maybe ideas for not needing temp files here in all cases? GH comment
+        temp_file = File.create(
+            client=self.client,
+            blocks=blocks,
+            tags=[Tag(kind=TagKind.GENERATION, name=GenerationTag.PROMPT_COMPLETION)],
+        )
+
+        try:
+            generation_task = self.plugin_instance.generate(
+                input_file_id=temp_file.id, options=kwargs
+            )
+            generation_task.wait()
+
+            for block in generation_task.output.blocks:
+                if block.mime_type == MimeTypes.STEAMSHIP_PLUGIN_CAPABILITIES:
+                    if logging.DEBUG >= logging.root.getEffectiveLevel():
+                        response = CapabilityPluginResponse.from_block(block)
+                        logging.debug(
+                            f"Plugin capability fulfillment:\n\n{pformat(response.json())}"
+                        )
+                    break
+            else:
+                if assert_capabilities:
+                    version_string = f"{self.plugin_instance.plugin_handle}, v.{self.plugin_instance.plugin_version_handle}"
+                    raise SteamshipError(
+                        f"Asserting capabilities are used, but capability response was not returned by plugin ({version_string})"
+                    )
+        finally:
+            # TODO (PR callout): It looks like current OpenAI impl might not do this?
+            temp_file.delete()
+
+        # TODO (PR callout): Use special output parser in the meantime to extract the tool to call from output block
+        #  tagged with new tags, _easy_ slot in with existing FunctionsBasedLLMAgent.
+        return generation_task.output.blocks
```

Comment:
In TagConstants, we have a new CAPABILITY_RESPONSE kind, name would be the name of the capability, and value would be per-capability.

e.g. in the case of Functions, return steamship.function-calling as the name, and "tool_name" as the value for which tool to call, and the block(s) that is(/are) tagged are the input to the tool.

Reply:
Do you have a clear code example of this somewhere?